gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireSnapshotsProcedure.java

```diff
@@ -0,0 +1,435 @@
+            new StructField(
+                OUTPUT_EXPIRED_MANIFEST_LISTS_COUNT, DataTypes.LongType, true, Metadata.empty()),
+            new StructField(
+                OUTPUT_EXPIRED_MANIFESTS_COUNT, DataTypes.LongType, true, Metadata.empty()),
+            new StructField(
+                OUTPUT_EXPIRED_FILES_LIST,
+                DataTypes.createArrayType(DataTypes.StringType),
+                true,
+                Metadata.empty())
+          });
+
+  public ExpireSnapshotsProcedure(TableCatalog currentCatalog) {
+    super(currentCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public String description() {
+    return String.format(
+        "Expires the Iceberg snapshots that are collected by '%s' procedure.",
+        IdentifyExpiredSnapshotsProcedure.PROCEDURE_NAME);
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow internalRow) {
+    String expiryBranchName = internalRow.getString(0);
+    String gcCatalogName = internalRow.getString(1);
+    String gcOutputBranchName = internalRow.getString(2);
+    String gcOutputTableIdentifier = internalRow.getString(3);
+    String gcCheckPointTableIdentifier = internalRow.getString(4);
+    Map<String, String> nessieClientConfig = new HashMap<>();
+    MapData map = internalRow.getMap(5);
+    for (int i = 0; i < map.numElements(); i++) {
+      nessieClientConfig.put(
+          map.keyArray().getUTF8String(i).toString(), map.valueArray().getUTF8String(i).toString());
+    }
+    String runId = !internalRow.isNullAt(6) ? internalRow.getString(6) : null;
+    boolean dryRun = !internalRow.isNullAt(7) && internalRow.getBoolean(7);
+
+    IdentifiedResultsRepo identifiedResultsRepo =
+        new IdentifiedResultsRepo(
+            spark(), gcCatalogName, gcOutputBranchName, gcOutputTableIdentifier);
+
+    runId = updateRunId(gcOutputTableIdentifier, runId, identifiedResultsRepo);
+
+    Map<String, List<Long>> expiredSnapshotsPerContentId =
+        getExpiredSnapshotsPerContentId(runId, identifiedResultsRepo);
+
+    Map<String, List<String>> metadataLocationsPerContentId =
+        getMetadataLocationsPerContentId(runId, identifiedResultsRepo);
+
+    List<InternalRow> outputRows = new ArrayList<>();
+    try (NessieApiV1 api = GCUtil.getApi(nessieClientConfig)) {
+      createGCBranch(expiryBranchName, api);
+      // get Nessie catalog
+      Catalog nessieCatalog = GCUtil.loadNessieCatalog(spark(), gcCatalogName, expiryBranchName);
+      String finalRunId = runId;
+      expiredSnapshotsPerContentId
+          .entrySet()
+          .forEach(
+              entry -> {
+                // Expiry Algorithm:
+                // A. For each content id, fetch ANY one of the content.
+                // B. Use the newly created expiry branch. On this branch,
```

Comment:
Why does this need a (temporary) branch, actually Nessie, at all?

Reply:
> Why does this need a (temporary) branch.
To create a dummy metadata table.  It may confuse existing users if created in other existing reference. Also, I need to drop that table as it will have reference to all the snapshots. Which will give incorrect results if not dropped.

yeah, other catalog works fine as well. Let me check it.

Reply:
Do we have to restrict Nessie user to configure some other catalog other than Nessie catalog? I find it odd. So, I used Nessie catalog. Any thoughts ?